File: src/hyperLiquid_ts_deprecated/helpers/telegram.ts
export async function sendTelegramMessage(chatId: string, token: string, message: string): Promise<void> {
    try {
        console.log("Sending Telegram message:", message);
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), 5000); // 5s timeout

        await fetch(`https://api.telegram.org/bot${token}/sendMessage`, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ chat_id: chatId, text: message }),
            signal: controller.signal
        });
        clearTimeout(timeoutId);
    } catch (err) {
        console.error("Failed to send Telegram message:", err);
    }
}

/**
 * Send trading notification with automatic message formatting
 * @param title - Notification title
 * @param symbol - Trading symbol
 * @param isBuy - Whether it's a buy order
 * @param price - Order price (accepts number or string)
 * @param stopLoss - Stop loss price (optional, accepts number or string)
 * @param pnl - PnL data (optional)
 * @param positionValue - Position value (optional)
 */
export async function sendNotification(
    title: string,
    symbol: string,
    isBuy: boolean,
    price: number | string,
    stopLoss?: number | string,
    positionValue?: number,
    pnl?: { netPnl: number; totalFees: number },
): Promise<void> {
    const isEnabled = process.env.TELEGRAM_ENABLED === 'true';

    if (!isEnabled) {
        console.log("Telegram notifications disabled");
        return;
    }

    const chatId = process.env.TELEGRAM_CHAT_ID;
    const token = process.env.TELEGRAM_BOT_TOKEN;

    if (!chatId || !token) {
        console.log("Telegram credentials not configured, skipping notification");
        return;
    }

    const priceNum = typeof price === 'string' ? parseFloat(price) : price;
    const stopLossNum = stopLoss ? (typeof stopLoss === 'string' ? parseFloat(stopLoss) : stopLoss) : undefined;

    const action = isBuy ? "🟢 BUY" : "🔴 SELL";
    let message = `${title}\n${action} ${symbol} @ ${priceNum}`;

    if (positionValue) {
        message += `\n💵 Position Value: $${positionValue.toFixed(2)}`;
    }

    if (stopLossNum) {
        message += `\nSL: ${stopLossNum}`;
    }

    if (pnl) {
        const pnlEmoji = pnl.netPnl >= 0 ? "💰" : "📉";
        message += `\n\n${pnlEmoji} Net PnL: $${pnl.netPnl.toFixed(2)}`;
        message += `\n💸 Fees: $${pnl.totalFees.toFixed(2)}`;
    }

    try {
        await sendTelegramMessage(chatId, token, message);
    } catch (error) {
        console.error("Failed to send notification:", error);
    }
}

File: src/hyperliquid_python/helpers/telegram.py
import os
import requests
from typing import Optional, Dict

def send_telegram_message(chat_id: str, token: str, message: str) -> None:
    try:
        url = f"https://api.telegram.org/bot{token}/sendMessage"
        payload = {"chat_id": chat_id, "text": message}
        requests.post(url, json=payload, timeout=5.0)
    except Exception as e:
        print(f"Failed to send Telegram message: {e}")

def send_notification(
    title: str,
    symbol: str,
    is_buy: bool,
    price: float,
    stop_loss: Optional[float] = None,
    position_value: Optional[float] = None,
    pnl: Optional[Dict[str, float]] = None
) -> None:
    is_enabled = os.environ.get('TELEGRAM_ENABLED') == 'true'
    
    if not is_enabled:
        print("Telegram notifications disabled")
        return
        
    from helpers.config_helpers import get_secret
    chat_id = os.environ.get('TELEGRAM_CHAT_ID')
    ssm_path = os.environ.get("SSM_TELEGRAM_TOKEN_PATH", "/hl/telegram_bot_token")
    token = get_secret('TELEGRAM_BOT_TOKEN', ssm_path)
    
    if not chat_id or not token:
        print("Telegram credentials not configured, skipping notification")
        return

    action = "🟢 BUY" if is_buy else "🔴 SELL"
    message = f"{title}\n{action} {symbol} @ {price}"
    
    if position_value:
        message += f"\n💵 Position Value: ${position_value:.2f}"
    
    if stop_loss is not None:
        message += f"\nSL: {stop_loss}"
        
    if pnl:
        # dict should have netPnl and totalFees
        net_pnl = pnl.get("netPnl", 0)
        total_fees = pnl.get("totalFees", 0)
        pnl_emoji = "💰" if net_pnl >= 0 else "📉"
        message += f"\n\n{pnl_emoji} Net PnL: ${net_pnl:.2f}"
        message += f"\n💸 Fees: ${total_fees:.2f}"

    try:
        send_telegram_message(chat_id, token, message)
    except Exception as e:
        print(f"Failed to send notification: {e}")
